README.md

```diff
@@ -1 +1,263 @@
+```
+
+Many text encodings of binary are inefficient and inconvenient. Others have opted to use canonicalization and a tag. This can be effective, but requires careful handling and signalling of the specific canonicalization method used.
+
+``` js
+const payload = canonicalize({"hello": "world", "count": 42})
+{payload: payload, sig: key.sign(sha256(payload))}
+```
+
+Directly signing over canonicalized introduces new problems: forced encoding and canonicalization attacks. By taking advantage of existing IPLD canonicalization, varsig aims to alleviate both.
+
+## 1.1 Forced Encoding
+
+Data must first be rendered to binary before it is signed. This means imposing an encoding. There is no standard way to include the encoding that some IPLD was encoded with other than a CID. In IPFS, CIDs imply a link, which can have implications for network access and storage. Further, generating a CID means producing a hash, which is then potentially rehashed by the cryptographic signature library.
+
+To remedy this, Varsig includes the encoding information used in production of the signature.
+
+## 1.2 Canonicalization Attacks
+
+Since IPLD is deterministically encoded, it can be tempting to rely on canonicalization at validation time, rather than rendering the IPLD to inline bytes or a CID and signing that. Since the original payload can be rederived from the output, this seems like a clean option:
+
+``` js
+// DAG-JSON
+{
+  "role": "user",
+  "links": [
+    {"/": "bafkreidb2q3ktgtlm5yio7buj3sypyghjtfh5ernsteqmakf4p2c5bwmyi"},
+    {"/": "bafkreic75ydg5vkw324oqkcmqltfvc3kivyngqkibjoysdwiilakh4z5fe"},
+    {"/": "bafkreiffdiz6raf46zrr3b2usufgz5fo44aggmocz4zappr6khhhljcdpy"}
+  ],
+  "sig": "8ufaS9w3CGN8cbQTUSoL1i7eaKiWLSXsD2LbZVmvM9zF"
+}
+```
+
+This opens the potential for [canonicalization attacks](https://soatok.blog/2021/07/30/canonicalization-attacks-against-macs-and-signatures/). Parsers are known to handle duplicate entries differently. IPLD needs to be serialized to a canonical form before checking the signature. Without careful handling, it is possible to fail to check if any additional fields have been added to the payload which will be parsed by the application. 
+
+> An object whose names are all unique is interoperable in the sense that all software implementations receiving that object will agree on the name-value mappings.  When the names within an object are not unique, the behavior of software that receives such an object is unpredictable.  Many implementations report the last name/value pair only.  Other implementations report an error or fail to parse the object, and some implementations report all of the name/value pairs, including duplicates.
+>
+> — [RFC8259](https://www.rfc-editor.org/rfc/rfc8259#page-10)
+
+``` json
+{
+  "role": "user",  // Parsed by an IPLD implementation
+  "role": "admin", // Malicious duplicate field, omitted by the IPLD parser, accpeted by the browser
```

Comment:
Typo: `accpeted` -> `accepted`

Reply:
Oh yeah, I guess the spell checker probably skips code blocks... hmmm...

Thanks!